ksqldb-engine/src/main/java/io/confluent/ksql/KsqlExecutionContext.java

```diff
@@ -125,6 +129,16 @@ TransientQueryMetadata executeQuery(
       ConfiguredStatement<Query> statement
   );
 
+  PullQueryResult executePullQuery(
+      ServiceContext serviceContext,
+      RoutingFilterFactory routingFilterFactory,
+      ConfiguredStatement<Query> statement,
+      Map<String, Object> requestProperties,
+      Optional<Boolean> isInternalRequest,
+      Optional<PullQueryExecutorMetrics> pullQueryMetrics,
+      RateLimiter rateLimiter
```

Comment:
I am not sure I am following what you mean.
The `rateLimiter` (after my last changes) is created once in the `KsqlRestApplication` and then used in `StreamedQueryResource` and `WsQueryEndpoint`.  After that, the logic of how it is applied hasn't changed from the `PullQueryExecutor`: it is used for every pull query. The same `rateLimiter` is being used for every pull query since it is created once.
I don't understand what you are saying that the design isn't safe. Alan's comment was about the `StreamedQueryResource` and `WsQueryEndpoint` having a different `rateLimiter` whereas we wanted to limit on their combined pull queries.

Reply:
I don't think it can be applied at the `KsqlRestApplication` as we don't know there whether a request is a pull query or not.